From d9b098e95ee6a2d4872287d43f8d4af943a1515e Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:50:32 -0400
Subject: [tools] perf/evlist: Add perf_evlist__to_front()

Message-id: <1408953210-26343-145-git-send-email-jolsa@redhat.com>
Patchwork-id: 88888
O-Subject: [PATCH RHEL7.1 BZ1133083 144/322] perf evlist: Add perf_evlist__to_front()
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit a025e4f0d8a92b38539d39b495b530015296b4d9
Author: Adrian Hunter <adrian.hunter@intel.com>
Date: Wed Dec 11 14:36:35 2013 +0200

description
===========
Add a function to move a selected event to the
front of the list.

This is needed because it is not possible
to use the PERF_EVENT_IOC_SET_OUTPUT IOCTL
from an Instruction Tracing event to a
non-Instruction Tracing event.  Thus the
Instruction Tracing event must come first.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/evlist.c b/tools/perf/util/evlist.c
index 2eb7378..0b31cee 100644
--- a/tools/perf/util/evlist.c
+++ b/tools/perf/util/evlist.c
@@ -1212,3 +1212,20 @@ int perf_evlist__strerror_open(struct perf_evlist *evlist __maybe_unused,
 
  return 0;
 }
+
+void perf_evlist__to_front(struct perf_evlist *evlist,
+      struct perf_evsel *move_evsel)
+{
+ struct perf_evsel *evsel, *n;
+ LIST_HEAD(move);
+
+ if (move_evsel == perf_evlist__first(evlist))
+  return;
+
+ list_for_each_entry_safe(evsel, n, &evlist->entries, node) {
+  if (evsel->leader == move_evsel->leader)
+   list_move_tail(&evsel->node, &move);
+ }
+
+ list_splice(&move, &evlist->entries);
+}
diff --git a/tools/perf/util/evlist.h b/tools/perf/util/evlist.h
index 8a04aae..9f64ede 100644
--- a/tools/perf/util/evlist.h
+++ b/tools/perf/util/evlist.h
@@ -194,5 +194,8 @@ static inline void perf_mmap__write_tail(struct perf_mmap *md,
 }
 
 bool perf_evlist__can_select_event(struct perf_evlist *evlist, const char *str);
+void perf_evlist__to_front(struct perf_evlist *evlist,
+      struct perf_evsel *move_evsel);
+
 
 #endif /* __PERF_EVLIST_H */
-- 
1.7.1